pgwire: add server.cockroach_cloud.max_client_connections_per_gateway cluster setting

alyshanjahani-crl opened the pull request:
This setting specifies a maximum number of external connections that a server can have open at any given time. External is defined as connections that are not root connections. In the future, we can have the definition of external be a configurable list of users.
<0 - Connections are unlimited (existing behavior)
=0 - External connections are disabled
>0 - External connections are limited
If a new connection would exceed this limit, an error message is returned as postgres: "cluster connections are limited" with the 53300 error code that corresponds to "too many connections".

This functionality is required for serverless.
Part of: https://cockroachlabs.atlassian.net/browse/CC-9288

Release note: None

alyshanjahani-crl committed:
pgwire: add server.cockroach_cloud.max_client_connections_per_gateway cluster setting

This setting specifies a maximum number of non root connections that a server
can have open at any given time.
<0 - Connections are unlimited (existing behavior)
=0 - Non root connections are disabled
>0 - Non root connections are limited
If a new connection would exceed this limit, an error message is
returned as postgres: "cluster connections are limited" with the 53300
error code that corresponds to "too many connections".

This functionality is required for serverless.
Part of: https://cockroachlabs.atlassian.net/browse/CC-9288

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100200)
<!-- Reviewable:end -->

knz reviewed (commented):
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @JeffSwenson, and @rafiss)

___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLC98EdOxgECD5Qha):*
Looks like this deserves a release note.
___
*[`pkg/sql/conn_executor.go` line 788 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLUXv6xuPY0RnWMil:-NRrLUXw3ODl4kDaK5zT:b2omaz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/conn_executor.go#L788)):*
> ```Go
> 
> // IncrementConnectionCount increases connectionCount by 1.
> // If isRoot is true then it also increases rootConnectionCount by 1.
> ```

I recommend you make two functions (root vs non-root) side by side. Bool arguments make reading the code harder.
___
*[`pkg/sql/pgwire/conn.go` line 721 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrMGpK7FBpFyhvRQC_:-NRrMGpK7FBpFyhvRQCa:b-qzuu98) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/pgwire/conn.go#L721)):*
> ```Go
> 		}
> 
> 		if retErr = c.checkMaxConnections(ctx, sqlServer); retErr != nil {
> ```

Q for @rafiss unrelated to this PR: should we check the max nr of connections before starting authentication? This way we can also contrain the amount of server-side authn work via the conn limit.
___
*[`pkg/sql/pgwire/server.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLxra3UEswCV5xIyh:-NRrLxra3UEswCV5xIyi:b8vq5cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/pgwire/server.go#L104)):*
> ```Go
> // Note(alyshan): This setting is not public.
> var maxNumExternalConnections = settings.RegisterIntSetting(
> 	settings.TenantReadOnly,
> ```

I don't understand why it's tenantreadonly. Why not tenantwritable like the other one?

(The cc control plane can still override as usual)


<!-- Sent from Reviewable.io -->

alyshanjahani-crl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @rafiss)

___
*[`-- commits` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLC98EdOxgECD5Qha):*
I felt that this change was not intended for external users, but just Cockroach Cloud.

> Do NOT write release note texts for internal changes, such as:
Functionality that is not accessible to, or intended for, external users (e.g., multi-tenant capabilities)
___
*[`pkg/sql/conn_executor.go` line 788 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLUXv6xuPY0RnWMil:-NRtIiFN0R9RiRjMgN1B:bn2tf3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/conn_executor.go#L788)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I recommend you make two functions (root vs non-root) side by side. Bool arguments make reading the code harder.
</blockquote></details>

In this case of an incoming root connection, are you imaging the caller make two calls `IncrementConnectionCount` and `IncrementRootConnectionCount` ?

Or one call to `IncrementRootConnectionCount` that increments both `connectionCount` and `rootConnectionCount`?
___
*[`pkg/sql/pgwire/server.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLxra3UEswCV5xIyh:-NRtG8wh7OULA9_-UVUa:bzi0flo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/pgwire/server.go#L104)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I don't understand why it's tenantreadonly. Why not tenantwritable like the other one?

(The cc control plane can still override as usual)
</blockquote></details>

I do not want serverless users to be able to discover and alter this setting on their own.

CC will set this setting for individual tenants through the system tenant.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @JeffSwenson, and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 788 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLUXv6xuPY0RnWMil:-NRvdNCf0dkHEw0mZrU4:b-6yr91k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/conn_executor.go#L788)):*
<details><summary><i>Previously, alyshanjahani-crl wrote…</i></summary><blockquote>

In this case of an incoming root connection, are you imaging the caller make two calls `IncrementConnectionCount` and `IncrementRootConnectionCount` ?

Or one call to `IncrementRootConnectionCount` that increments both `connectionCount` and `rootConnectionCount`?
</blockquote></details>

the latter.
___
*[`pkg/sql/pgwire/server.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLxra3UEswCV5xIyh:-NRvdPIdC52zQBomD9dR:b-1j3u5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/pgwire/server.go#L104)):*
> I do not want serverless users to be able to discover and alter this setting on their own.

They would not be able to alter it if CC orchestration sets it via an override. Overrides cannot be altered by users.

IMHO there is no good reason to make it TenantReadOnly since it's the same type as the other one just above.


<!-- Sent from Reviewable.io -->

alyshanjahani-crl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 788 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLUXv6xuPY0RnWMil:-NS6tabIAHTFrmYZtPfC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/conn_executor.go#L788)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

the latter.
</blockquote></details>

Done.
___
*[`pkg/sql/pgwire/server.go` line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrLxra3UEswCV5xIyh:-NS6tcdU3aCjOVourFDn:b-yycbaf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/pgwire/server.go#L104)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

> I do not want serverless users to be able to discover and alter this setting on their own.

They would not be able to alter it if CC orchestration sets it via an override. Overrides cannot be altered by users.

IMHO there is no good reason to make it TenantReadOnly since it's the same type as the other one just above.

</blockquote></details>

Oh gotchya, that makes sense to me now.

Changed to `TenantWritable`


<!-- Sent from Reviewable.io -->

alyshanjahani-crl committed:
pgwire: add server.cockroach_cloud.max_client_connections_per_gateway_reason cluster setting

This setting can be used to customize the error message returned when
connections are denied due to a limit specified via
server.cockroach_cloud.max_client_connections_per_gateway.

This functionality is required for serverless. Being able to indicate to the
user why a limit is placed on their number of connections will support a better
UX. Specifically, indicating to the user which resource limits they have hit.
Part of: https://cockroachlabs.atlassian.net/browse/CC-9288

Release note: None

rafiss reviewed (changes requested):
just had minor comments

i was wondering - is there a risk for confusion if this new setting and the old `max_per_gateway` setting are used at the same time? does one take precedence?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-:-NSCY-LjDG24q_QAAP2H:b-t4u94x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl, @JeffSwenson, and @knz)

___
*[`pkg/sql/conn_executor.go` line 804 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NSCWvT5CrGTHSl2NPp-:-NSCWvT6DsqvycUdX86n:b3q3pdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/bce5c022c91254a247b9d0aa9f2a423dd43456df/pkg/sql/conn_executor.go#L804)):*
> ```Go
> 
> // DecrementConnectionCount decreases connectionCount by 1.
> // If isRoot is true then it also decreases rootConnectionCount by 1.
> ```

nit: this comment seems outdated
___
*[`pkg/sql/pgwire/conn.go` line 721 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NRrMGpK7FBpFyhvRQC_:-NSCX10nC2LduzECBk7I:b-ky7mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a05ec83d7e4227c71b37f2539d8671ffbf76f5aa/pkg/sql/pgwire/conn.go#L721)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Q for @rafiss unrelated to this PR: should we check the max nr of connections before starting authentication? This way we can also contrain the amount of server-side authn work via the conn limit.
</blockquote></details>

i believe we'd discussed this before, and decided no. this would allow someone who has no access to the cluster to spam connection attempts, and thereby lock out legitimate users from ever being able to begin authentication attempts.
___
*[`pkg/sql/pgwire/server.go` line 100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NSCXVP46C8jDTedgvFp:-NSCXVP46C8jDTedgvFq:bdqxwkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bce5c022c91254a247b9d0aa9f2a423dd43456df/pkg/sql/pgwire/server.go#L100)):*
> ```Go
> // Note(alyshan): One might suggest this cluster setting be named server.max_non_root_connections_per_gateway
> // as that reflects its current behaviour of excluding root connections from being limited. However, I chose
> // to use the term "external" so that this setting can be extended to exclude connections from an arbitrary
> ```

the name "external connection" is unfortunate here. that already has a definition in our product https://www.cockroachlabs.com/docs/stable/create-external-connection.html

can we just name this `maxNumNonRootConnections`? and name the setting `server.max_non_root_connections_per_gateway`?

can you say more about how you imagine extending the setting?


<!-- Sent from Reviewable.io -->

alyshanjahani-crl reviewed (commented):
> does one take precedence?

The one im adding `maxNonRoot` is checked first since it is more restrictive. But they won't interfere with each other.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-:-NSChPE3E037_9sgRCpm:bamu6mj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @rafiss)

___
*[`pkg/sql/pgwire/server.go` line 100 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NSCXVP46C8jDTedgvFp:-NSChtYtDN9mHV-YViuG:b-ffoio1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bce5c022c91254a247b9d0aa9f2a423dd43456df/pkg/sql/pgwire/server.go#L100)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the name "external connection" is unfortunate here. that already has a definition in our product https://www.cockroachlabs.com/docs/stable/create-external-connection.html

can we just name this `maxNumNonRootConnections`? and name the setting `server.max_non_root_connections_per_gateway`?

can you say more about how you imagine extending the setting?
</blockquote></details>

Currently when CockroachCloud performs operations on serverless clusters (like the console page listing sql users, creating databases etc, doing a CRDB version upgrade etc..) they connect as root.

The intention behind this cluster setting is to let an administrator access the cluster but limit other users. Currently the `root` user is this administrator. But that might not always be the case, we might have more fine grained access control and so i was imagining extending this setting one day with another setting where one can provide which users should be exempted / treated like `root` in this case.

I wanted to avoid a renaming of cluster setting in the future, and so i thought of `maxExternalConnections`.

I suppose i would be fine with `server.max_non_root_connections_per_gateway`, and if we ever needed to extend the functionality of the setting as i described above, we would make it explicit/clear in the description of this setting that it can permit more than just `root` connections via using an additional setting.


<!-- Sent from Reviewable.io -->

jeffswenson reviewed (commented):
From a high level architecture standpoint, I agree that we should implement this in the sql server instead of the sql proxy. The one downside is adding a setting to the sql server means we need to support it O(forever) because removing the setting will be a breaking change. That means we need to take some time and make sure the setting makes sense outside of the context of CC. We also need to think about how throttling will interact with a user that has configured the setting themselves.

As for the design, I'm hesitant to add a new setting that has a behavior very similar to an existing setting, because there are already a lot of settings and I don't want to make things more confusing.

Here are a few ideas:
1. We add a setting which defines which users are exempt from checkMaxConnections. This feature would be more useful for non CC users and we can override the two settings together 
2. We stick with the somewhat hacky sql proxy change, because we can remove the feature if we no longer want to support it the future. E.g. once we have a GUI sql shell, maybe we limit all connections at the sql proxy and only allow connections through the GUI shell.

alyshanjahani-crl reviewed (commented):
I spoke w. Jeff offline, the setting being private/reserved actually addresses his concerns. With the caveat of including in the description of the setting a disclaimer/warning that this is intended to be used by Cockroach Cloud, as well as a comment in the usage of the setting in `checkMaxConnections` 

And on a similar note, a request to include `cloud` somewhere in the name of the setting.

Before I go ahead and change the name of the setting again,  can we agree upon a name here first?
Is there any rules around what can be the prefix of the cluster setting name @rafiss ? for ex. instead of `server.max_conn...` could we do `cloud.max_conn...` ?
If we can that my preferred cluster setting name is `cloud.max_connections_per_gateway`  with a description like so
```
This setting is intended to be used by Cockroach Cloud for limiting connections to serverless clusters.
The maximum number of connections per gateway allowed at a given time
(note: this will only limit future connection attempts and will not affect already established connections).
Negative values result in unlimited number of connections. Cockroach cloud internal users (including root user) are not affected by this limit."
```

Alternatively: `server.max_non_cloud_connections_per_gateway`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-:-NSHfqWo013ITXQ-_mRQ:b-vzbjq7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss commented:
what about `server.max_cockroach_cloud_connections_per_gateway`. (i want to use `cockroach_cloud` instead of `cloud`, since `cloud` is used in other parts of CRDB code to refer to the idea of connecting to resources like S3 or GCE)

the general convention is that the first part of the cluster setting name relates to the code/component that is being configured.

alyshanjahani-crl reviewed (commented):
ack, i've used `server.max_non_cockroach_cloud_connections_per_gateway` . It is a mouthful, but to re-iterate, this is a setting that is private, only intended for cockroach cloud, and will possibly be removed in the future when we have a better way to limit connections to serverless clusters.

TFTR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-:-NSNHdwBBJ3vuodyYh8g:b1utrq0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @rafiss)



<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl and @knz)

___
*[`pkg/sql/pgwire/server.go` line 104 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NSNMbQ49z_CoikqPIDc:-NSNMbQ49z_CoikqPIDd:bmfazaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc20ee6762f99cd1ecf6a4495647a8841dd8fb07/pkg/sql/pgwire/server.go#L104)):*
> ```Go
> var maxNumNonRootConnections = settings.RegisterIntSetting(
> 	settings.TenantWritable,
> 	"server.max_non_cockroach_cloud_connections_per_gateway",
> ```

nit: it should be `server.max_cockroach_cloud_connections_per_gateway`

sorry for the review churn! feel free to merge after fixing this.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl and @knz)

___
*[`pkg/sql/pgwire/server.go` line 104 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NSNMbQ49z_CoikqPIDc:-NSNMlmaDXWL4UysWRNc:b2mqdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc20ee6762f99cd1ecf6a4495647a8841dd8fb07/pkg/sql/pgwire/server.go#L104)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: it should be `server.max_cockroach_cloud_connections_per_gateway`

sorry for the review churn! feel free to merge after fixing this.
</blockquote></details>

well, my idea was "cockroach cloud connection" refers to someone connecting to cockroach cloud

but i see in your description that you are thinking of it differently maybe?


<!-- Sent from Reviewable.io -->

JeffSwenson commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl and @knz)
> 
> _[`pkg/sql/pgwire/server.go` line 104 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NSNMbQ49z_CoikqPIDc:-NSNMlmaDXWL4UysWRNc:b2mqdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc20ee6762f99cd1ecf6a4495647a8841dd8fb07/pkg/sql/pgwire/server.go#L104)):_
> 
> Previously, rafiss (Rafi Shamim) wrote…
> well, my idea was "cockroach cloud connection" refers to someone connecting to cockroach cloud
> 
> but i see in your description that you are thinking of it differently maybe?

The intent of the setting is to differentiate between connections from the CC control plane vs connections from the CC user. What about the name `server.cockroach_cloud_client_connection_limit`?

JeffSwenson commented:
> > > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100200)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @alyshanjahani-crl and @knz)
> > 
> > 
> > _[`pkg/sql/pgwire/server.go` line 104 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/100200#-NSNMbQ49z_CoikqPIDc:-NSNMlmaDXWL4UysWRNc:b2mqdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc20ee6762f99cd1ecf6a4495647a8841dd8fb07/pkg/sql/pgwire/server.go#L104)):_
> > Previously, rafiss (Rafi Shamim) wrote…
> > well, my idea was "cockroach cloud connection" refers to someone connecting to cockroach cloud
> > but i see in your description that you are thinking of it differently maybe?
> 
> The intent of the setting is to differentiate between connections from the CC control plane vs connections from the CC user. What about the name `server.cockroach_cloud_client_connection_limit`?

Or maybe
`server.cockroach_cloud.max_client_connections`.

rafiss commented:
`server.cockroach_cloud.max_client_connections_per_gateway` sounds like a good one to me. (the `per_gateway` part is important, i think)

rafiss commented on pkg/sql/pgwire/server.go:
```diff
@@ -82,7 +82,11 @@ var logSessionAuth = settings.RegisterBoolSetting(
 	false).WithPublic()
 
-var maxNumConnections = settings.RegisterIntSetting(
+// TODO(alyshan): This setting is enforcing max number of connections with superusers not being affected by
```
good call on the TODO. we should do this one:
> update the description to say "the maximum number of connections per gateway ... Superusers are not affected by this limit"

(our team can do this in a later PR if that's easier)

alyshanjahani-crl replied:
In spirit of breather week:

https://github.com/cockroachdb/cockroach/pull/101177

rafiss commented on pkg/sql/pgwire/server.go:
```diff
@@ -91,6 +95,34 @@ var maxNumConnections = settings.RegisterIntSetting(
+)
+
+// maxNumNonRootConnectionsReason is used to supplement the error message for connections that denied due to
+// server.max_non_root_connections_per_gateway.
```
super nit: this comment refers to the old name

alyshanjahani-crl replied:
thanks for catching, fixed

rafiss reviewed (approved):
lgtm! thanks for the attention to detail

alyshanjahani-crl commented:
bors r+

alyshanjahani-crl commented:
blathers backport 22.2 23.1
